src/Components/Web/test/Forms/InputRadioTest.cs

```diff
@@ -0,0 +1,161 @@
+            };
+
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(() => RenderAndGetTestInputComponentAsync(rootComponent));
+            Assert.Contains($"requires either an explicit string attribute 'name' or an ancestor", ex.Message);
+        }
+
+        [Fact]
+        public async Task GeneratesNameGuidWhenInvalidNameSuppliedWithGroup()
+        {
+            var model = new TestModel();
+            var rootComponent = new TestInputRadioHostComponent<TestEnum>
+            {
+                EditContext = new EditContext(model),
+                InnerContent = RadioButtonsWithGroup(null, () => model.TestEnum)
+            };
+
+            var inputRadioComponents = await RenderAndGetTestInputComponentAsync(rootComponent);
+
+            Assert.All(inputRadioComponents, inputRadio => Assert.True(Guid.TryParseExact(inputRadio.GroupName, "N", out _)));
+        }
+
+        [Fact]
+        public async Task NameAttributeExistsWhenValidNameSupplied_WithoutGroup()
+        {
+            var groupName = "group";
+            var model = new TestModel();
+            var rootComponent = new TestInputRadioHostComponent<TestEnum>
+            {
+                EditContext = new EditContext(model),
+                InnerContent = RadioButtonsWithoutGroup(groupName, () => model.TestEnum)
+            };
+
+            var inputRadioComponents = await RenderAndGetTestInputComponentAsync(rootComponent);
+
+            Assert.All(inputRadioComponents, inputRadio => Assert.Equal(groupName, inputRadio.GroupName));
+        }
+
+        [Fact]
+        public async Task NameAttributeExistsWhenValidNameSupplied_WithGroup()
+        {
+            var groupName = "group";
+            var model = new TestModel();
+            var rootComponent = new TestInputRadioHostComponent<TestEnum>
+            {
+                EditContext = new EditContext(model),
+                InnerContent = RadioButtonsWithGroup(groupName, () => model.TestEnum)
+            };
+
+            var inputRadioComponents = await RenderAndGetTestInputComponentAsync(rootComponent);
+
+            Assert.All(inputRadioComponents, inputRadio => Assert.Equal(groupName, inputRadio.GroupName));
+        }
+
+        private delegate RenderFragment InputRadioGenerator(string name, Expression<Func<TestEnum>> valueExpression);
+
+        private static readonly InputRadioGenerator RadioButtonsWithoutGroup = (name, valueExpression) => (builder) =>
+        {
+            int sequence = 0;
+
+            foreach (var selectedValue in (TestEnum[])Enum.GetValues(typeof(TestEnum)))
+            {
+                builder.OpenComponent<TestInputRadio>(sequence++);
+                builder.AddAttribute(sequence++, "name", name);
+                builder.AddAttribute(sequence++, "SelectedValue", selectedValue);
+                builder.AddAttribute(sequence++, "ValueExpression", valueExpression);
```

Comment:
Would it be OK to avoid using this `sequence++` pattern here? I know it might not cause any particular problems in this test case, but people copy the patterns they see in our sources, and I've been trying to be consistent with the community about [why sequence numbers can't be generated dynamically](https://gist.github.com/SteveSandersonMS/ec232992c2446ab9a0059dd0fbc5d0c3).

In this case, you can hard-code the sequence numbers with any incrementing value that corresponds to the source-code line sequence. For example:

```suggestion
                builder.OpenComponent<TestInputRadio>(0);
                builder.AddAttribute(1, "name", name);
                builder.AddAttribute(2, "SelectedValue", selectedValue);
                builder.AddAttribute(3, "ValueExpression", valueExpression);
```

It's neither necessary nor desirable for the sequence numbers to continue increasing on subsequent loops of iteration. They are meant to reset back to a lower number on each loop, because that's how diffing knows what's going on.

Reply:
Same in the other cases below.